functionary/train/train_lora.py

```diff
@@ -0,0 +1,396 @@
+import json
+import math
+import pathlib
+import random
+from dataclasses import dataclass, field
+from typing import Dict, Optional
+
+import torch
+import torch.distributed
+from torch.nn import CrossEntropyLoss
+import sys
+import os
+
+os.environ["WANDB_LOG_MODEL"] = "all"
+
+sys.path.append(os.path.join(os.path.dirname(__file__), "../.."))
+from functionary.prompt import EndToken
+from functionary.train.custom_datasets import CustomDataset
+
+from deepspeed import zero
+from deepspeed.runtime.zero.partition_parameters import ZeroParamStatus
+
+import transformers
+from transformers import LlamaTokenizer, Trainer, BitsAndBytesConfig, deepspeed
+from transformers import LlamaForCausalLM
+from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
+import bitsandbytes as bnb
+
+LOCAL_RANK = int(os.getenv("LOCAL_RANK", "0"))
+
+
+def print_rank0(*arg):
+    if LOCAL_RANK == 0:
+        print(*arg)
+
+
+@dataclass
+class ModelArguments:
+    model_name_or_path: Optional[str] = field(default="meta-llama/Llama-2-7b-hf")
+
+
+@dataclass
+class DataArguments:
+    train_data_path: str = field(default=None, metadata={"help": "Path to the training data."})
+    eval_data_path: str = field(default=None, metadata={"help": "Path to the eval data."})
+
+
+@dataclass
+class TrainingArguments(transformers.TrainingArguments):
+    cache_dir: Optional[str] = field(default=None)
+    model_max_length: int = field(
+        default=4096,
+        metadata={"help": "Maximum sequence length. Sequences will be right padded (and possibly truncated)."},
+    )
+
+
+def safe_save_model_for_hf_trainer(trainer: transformers.Trainer, output_dir: str):
+    """Collects the state dict and dump to disk."""
+    state_dict = trainer.model.state_dict()
+    if trainer.args.should_save:
+        cpu_state_dict = {key: value.cpu() for key, value in state_dict.items()}
+        del state_dict
+        trainer._save(output_dir, state_dict=cpu_state_dict)  # noqa
+
+
+@dataclass
+class LoraArguments:
+    lora_r: int = 16
+    lora_alpha: int = 64
+    lora_dropout: float = 0.1
+    lora_target_modules: str = "all"  # all for all linear; "q_proj v_proj"
+    lora_weight_path: str = ""
+    lora_bias: str = "none"
+    q_lora: bool = False
+
+
+def initialize_tokenizer(
```

Comment:
Can you use the same tokenizer initialization instead of duplicating here

Reply:
what do you mean by using the same tokenizer initialization?

Reply:
I meant, we already have initialize_tokenizer in the train.py. So we can import it instead